src/librbd/io/ImageRequestWQ.cc

```diff
@@ -787,26 +919,90 @@ void *ImageRequestWQ<I>::_void_dequeue() {
 }
 
 template <typename I>
-void ImageRequestWQ<I>::process(ImageDispatchSpec<I> *req) {
+void ImageRequestWQ<I>::process_io(ImageDispatchSpec<I> *req,
+    bool non_blocking_io) {
   CephContext *cct = m_image_ctx.cct;
   ldout(cct, 20) << "ictx=" << &m_image_ctx << ", "
                  << "req=" << req << dendl;
 
+  //extents are invalidated after the request is sent
+  //so gather them ahead of that
+  const auto& extents = req->get_image_extents();
+  bool write_op = req->is_write_op();
+  uint64_t tid = req->get_tid();
+  uint64_t offset;
+  uint64_t length;
+
+  if (write_op) {
+    std::lock_guard locker{m_lock};
+    offset = extents.size() ? extents.front().first : 0;
+    length = extents.size() ? extents.front().second : 0;
+    bool blocked = block_overlapping_io(&m_in_flight_extents, offset, length);
+    if (blocked) {
+      ldout(cct, 20) << "blocking overlapping IO: " << "ictx="
+                     << &m_image_ctx << ", "
+                     << "off=" <<  offset << ", len=" << length << dendl;
+      m_blocked_ios.push_back(req);
+      --m_in_flight_ios;
+      return;
+    }
+  }
+
   req->send();
 
-  finish_queued_io(req);
-  if (req->is_write_op()) {
-    finish_in_flight_write();
+  if (write_op) {
+    if (non_blocking_io) {
+      finish_in_flight_write();
+    }
+    unblock_overlapping_io(offset, length, tid);
+    unblock_flushes(tid);
   }
+}
+
+template <typename I>
+void ImageRequestWQ<I>::process(ImageDispatchSpec<I> *req) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "ictx=" << &m_image_ctx << ", "
+                 << "req=" << req << dendl;
+
+  bool write_op = req->is_write_op();
+  process_io(req, true);
+
+  finish_queued_io(write_op);
   delete req;
```

Comment:
Nit: move this to `process_io` or else you will have a memory leak on the blocking IO path

Reply:
Done